assignment4/Q1.cpp

```diff
@@ -0,0 +1,126 @@
+#include <iostream>
+#include <vector>
+#include <set>
+
+using namespace std;
+//iterating two times - first to copy the board
+//not sure of its correctness
+//probably something is missing but I do not know yet
+
+struct Point {
+    int x;
+    int y;
+    char val;
+    int set_no;
+};
+
+//returns the set whom point belongs to
+int findSet(struct Point point) {
+    return point.set_no;
+}
+
+void unionSets(set<int> &sets, int set_to_erase, int set_to_stay) {
+
+    sets.erase(set_to_erase);
+    return;
+}
+
+int count_islands(vector <vector<char> > board) {
+
+    int height = board.size();
+    int width = board[0].size();
+    int set_no = 0;
+    set<int> sets; //for storing sets' numbers
+
+    //initialize board with guards
+    struct Point template_point;
+    template_point.x = template_point.y = -1;
+    template_point.val = 'F';
+    template_point.set_no = -1;
+
+    vector< vector<struct Point> > board_with_guards(height+1, vector<struct Point>(width+1, template_point));
+    for(int i = 1; i < height + 1; i++){
+        for(int j = 1; j < width + 1; j++){
+            struct Point point;
+            point.x = i - 1;
+            point.y = j - 1;
+            point.val = board[i-1][j-1];
+            board_with_guards[i][j] = point;
+        }
+    }
+
+    //set numbers on board are just for joining
+    //their main point is to add to set sets and know which one to delete
+
+    for(int i = 1; i < height + 1; i++){
+        for(int j = 1; j < width + 1; j++) {
+            if(board_with_guards[i][j].val == 'F')
+                continue;
+            else {
+                board_with_guards[i][j].set_no = ++set_no;
+                sets.insert(set_no);
+
+                //join with upper neighbour if 'T'
+                struct Point up = board_with_guards[i-1][j];
+                if(up.val == 'T') {
+                    int upperset_no = findSet(up);
+                    unionSets(sets, set_no, upperset_no);
+                    board_with_guards[i][j].set_no = upperset_no;
+                }
+
+                //join with left neighbour if 'T'
+                struct Point left = board_with_guards[i][j-1];
+                if(left.val == 'T' && up.val != 'T') {
+                    int leftset_no = findSet(left);
+                    unionSets(sets, board_with_guards[i][j].set_no, leftset_no);
+                    board_with_guards[i][j].set_no = leftset_no;
+                }
+                if(left.val == 'T' && up.val == 'T') {
+                    int upperset_no = findSet(up);
+                    int leftset_no = board_with_guards[i][j-1].set_no;
+                    if(leftset_no != upperset_no)
+                        unionSets(sets, leftset_no, upperset_no);
+                    board_with_guards[i][j-1].set_no = upperset_no;
```

Comment:
Note that there is no need for this line. The cell from the left already belongs to an "invalid" island, so it can keep its original index.
See the following example:
```
FFT
TTT
```
With this line, the island indices would be:
```
FF1
211
```
And without it, they would be
```
FF1
221
```
In both ways, the leftmost island will keep its index 2. And it is ok, since this island is already invalidated, so it is ok of more tiles will be added to it. So there is no harm in keeping `[i][j-1]` cell with its original index.